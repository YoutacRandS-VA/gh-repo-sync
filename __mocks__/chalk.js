const chalk = jest.genMockFromModule("chalk");

chalk.cyan = jest.fn((str) => str);
chalk.blue = jest.fn((str) => str);
chalk.green = jest.fn((str) => str);
chalk.yellow = jest.fn((str) => str);
chalk.red = jest.fn((str) => str);

module.exports = chalk;
