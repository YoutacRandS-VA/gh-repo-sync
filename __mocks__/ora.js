const result = {
  start: jest.fn(),
  stop: jest.fn(),
  succeed: jest.fn(),
  fail: jest.fn(),
};
const ora = jest.fn(() => result);

module.exports = ora;
