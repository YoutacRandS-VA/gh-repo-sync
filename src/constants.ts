import { readJsonSync } from "fs-extra";
import { resolve } from "path";

const _pkg = readJsonSync(resolve(__dirname, "../package.json"));

const _toolName = _pkg.name;
const _prefix = `[${_toolName}]`;

export const pkg = _pkg;
export const toolName = _toolName;
export const prefix = _prefix;
