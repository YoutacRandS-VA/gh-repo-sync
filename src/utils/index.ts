export { default as runCommand } from "./run-command";
export { default as sync } from "./sync";
export { default as renderError } from "./render-error";
export { default as renderInfo } from "./render-info";
export { default as renderSucceed } from "./render-succeed";
