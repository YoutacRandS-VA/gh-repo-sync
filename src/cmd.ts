import { program } from "commander";
import { ifExistGH, ifLoginGH, renderError, renderInfo, sync } from "./utils";
import { pkg } from "./constants";

program
  .version(pkg.version, "-v, --version")
  .usage("[options]")
  .option("-s, --string <your string>", "Set your string")
  .parse(process.argv);

export async function run() {
  process.stdout.write(renderInfo("Processing...\n"));

  const pendingRepos = [
    "hexh250786313/coc-symbol-line",
    "hexh250786313/coc-yank",
    "hexh250786313/coc-git",
  ];

  try {
    await ifExistGH();
    await ifLoginGH();
    Promise.all(
      pendingRepos.map(async (repo) => {
        await sync(repo);
      })
    );
    process.stdout.write("\n");
  } catch (e: any) {
    if (e.message) {
      process.stdout.write(renderError(e as string) + "\n");
    }
  }
}
