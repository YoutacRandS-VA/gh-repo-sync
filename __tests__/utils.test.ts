import chalk from "chalk";
import commandExists from "command-exists";
import ora from "ora";
import { prefix } from "../src/constants";
import {
  ifExistGH,
  ifLoginGH,
  renderError,
  renderInfo,
  renderSucceed,
  runCommand,
  spinner,
  sync,
} from "../src/utils";

const spyWrite = jest
  .spyOn(global.process.stdout, "write")
  .mockImplementation();

describe("Unit test for runCommand", () => {
  it("Should throw on command error", async () => {
    let err: unknown;
    try {
      await runCommand("command_not_exists", { cwd: __dirname });
    } catch (e) {
      err = e;
    }
    expect(err).toBeDefined();
  });

  it("Should run command with timeout", async () => {
    let err: unknown;
    try {
      await runCommand("sleep 2", { cwd: __dirname }, 0.01);
    } catch (e) {
      err = e;
    }
    expect(err).toBeDefined();
  });
});

describe("Unit test for render utils", () => {
  it("Should call chalk with correct arguments", () => {
    const error = "error";
    const info = "info";
    const succeed = "succeed";

    renderInfo(info);
    expect(chalk.cyan).toHaveBeenLastCalledWith(`${prefix} ${info}`);

    renderSucceed(succeed);
    expect(chalk.green).toHaveBeenLastCalledWith(`${prefix} ${succeed}`);

    renderError(error);
    expect(chalk.red).toHaveBeenLastCalledWith(`${prefix} ${error}`);
  });
});

describe("Unit test for spinner", () => {
  let oraOra: ora.Ora;
  let calledAmount: number;
  const testStr = "test text";

  beforeEach(() => {
    calledAmount = (ora().start as any).mock.calls.length;
    oraOra = spinner(testStr);
  });

  it("Should start an ora.Ora process", () => {
    expect(ora().start).toHaveBeenCalledTimes(calledAmount + 1);
  });

  it("Should call with correct argument", () => {
    expect(chalk.cyan).toHaveBeenLastCalledWith(testStr);
    expect(ora).toHaveBeenLastCalledWith({
      color: "cyan",
      text: chalk.cyan(testStr),
    });
  });

  it("Should return current ora.Ora instance", () => {
    expect(typeof oraOra.stop).toBe("function");
    expect(typeof oraOra.fail).toBe("function");
    expect(typeof oraOra.succeed).toBe("function");
    expect(typeof oraOra.start).toBe("function");
  });
});

describe("Unit test for ifExistGH", () => {
  let calledAmount: number;

  beforeEach(async () => {
    try {
      await ifExistGH();
    } catch (e) {
      /* handle error */
    }
  });

  it("Should call with correct arguments", () => {
    expect(commandExists).toHaveBeenLastCalledWith("gh");
  });

  it("Should print correct error message", () => {
    calledAmount = (process.stdout.write as any).mock.calls.length;
    expect(
      (process.stdout.write as any).mock.calls[calledAmount - 1][0]
    ).toMatchSnapshot();
  });
});

describe("Unit test for ifLoginGH", () => {
  let calledAmount: number;

  beforeEach(async () => {
    try {
      await ifLoginGH();
    } catch (e) {
      /* handle error */
    }
  });

  // @todo:
  // it("Should call with correct arguments", () => {
  // expect(runCommand).toHaveBeenLastCalledWith(`gh auth status`);
  // });

  it("Should print correct error message", () => {
    calledAmount = (process.stdout.write as any).mock.calls.length;
    expect(
      (process.stdout.write as any).mock.calls[calledAmount - 1][0]
    ).toMatchSnapshot();
    spyWrite.mockRestore();
  });
});

describe("Unit test for sync", () => {
  let calledAmount: number;
  beforeEach(async () => {
    calledAmount = (ora().start as any).mock.calls.length;
    try {
      await sync("");
    } catch (err) {
      // No handler needed
    }
  });

  it("Should start an ora.Ora process", () => {
    expect(ora().start).toHaveBeenCalledTimes(calledAmount + 1);
  });
});
