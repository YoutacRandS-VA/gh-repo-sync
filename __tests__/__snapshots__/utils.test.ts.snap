// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for ifExistGH Should print correct error message 1`] = `
"
[gh-repo-sync] Error: Command failed: gh: command not found

Please make sure you have github-cli installed: https://github.com/cli/cli#installation

"
`;

exports[`Unit test for ifLoginGH Should print correct error message 1`] = `
"
You are not logged into any GitHub hosts. Run \`gh auth login\` to authenticate.

"
`;
